Retain task history after splitting; fixes #156

ethan-nelson opened the pull request:
Based on #156, #612, and personal history, retaining history after a tile split is important. This pull request accomplishes history retention by copying over the task states, task locks, and task comments to the newly generated tiles. 

One main issue when doing this is that the time stamps of history are lost when the initializations of `TaskState` et al automatically assign the time when those new rows are initialized. To counter this, I have modified the models for each of the three history components to include an optional `date` keyword argument. My one concern is whether this introduces any major inefficiencies.

If this is indeed an issue, we can instead create a temporary variable in each of the three history object loops (L360-367), set the `date` attribute of that variable to the given time stamp, and then append that variable to the new tile history objects. I had previously tried implementing something like `t.states.append(TaskState(state.user, state.state).setattr('date',state.date)` but received an error. (There may be a slicker sqlalchemy way of modifying the attribute in one line that could also be used instead of this temporary variable idea.)

Moving along, now that we are assigning the history the correct time stamps, we run into another issue when one `TaskLock` and `TaskState` are automatically generated for the new task initialization. Again, the time of initialization is assigned to the time stamp for those actions. This causes the history to be out of order and incorrect when the task history is loaded on the task screen. So, we clear the state and lock values of each new task and instead populate it with the old task's while including the state and lock values from the original initialization of the parent tile.

Next, there should be some indication that the tile was split and who split it. We take advantage of the fact that when the task history is loaded on the task screen, the initial state and lock values from the original initialization are popped out of the listing. _That means the only remaining `state_ready` values that will exist in the history are those that we assign after a tile has been successfully split!_. So, the task history template adds that a tile has been split when a `state_ready` event remains in the history. 

Lastly, because we have modified the tile splitting to now include old history, the `check_for_updates2` function in the tests has to be updated to account for this as it tests the activity after a tile split.

Long explanation with some convolution :smile: comments and feedback encouraged.

ethan-nelson commented:
![example history](https://cloud.githubusercontent.com/assets/8998918/14626837/9e37534e-05b4-11e6-9e6c-d4f10bfada53.PNG)

pgiraud commented:
@ethan-nelson I did not go deeply on your code though I read your explanations. What comes to my mind at first is that it doesn't seem simple. Very naively I didn't think that it would take that much of efforts.
What looks strange to me is that history from the parent tile gets copied. It looks like duplicated information. And I don't like when things are duplicated.

I agree with you, the information about splitting should be visible somewhere.

Very naively again, did you think about storing the information about a potential "parent" tile along with having a new "split" state?
Here's what I imagine:
 - user splits task,
 - task state is set to "split" with date and user information stored,
 - new generated tiles (or at least the new "ready" states for those tiles) get a parent id corresponding to the parent tile (or parent state).

When displaying history for a tile we can then check if task were created after a split (parent is not None) and if appropriate go up in the parent history.

ethan-nelson commented:
Interesting methodology. I generally tend to avoid altering DB schema, so that is why i went with this approach :). But I like this idea of adding a parent_id so that when we query history of a tile, we look for events corresponding to that project/tile + project/tile_parent (with the latter having its own project/tile_parent and so on). I'll take a stab with this different method this weekend.

pgiraud commented:
Don't hesitate to change the model if this is not for a complete rewrite. It's `alembic`'s job to ensure that do it safely. If you need help with `alembic` don't hesitate to ask.

ethan-nelson commented:
So this definitely isn't complete, but my lack of web dev knowledge is starting to show in here in that I'm a bit stuck.

I added a relationship so that when a tile is split, the new tiles can reference their parent tile easily. But when it comes to loading the history, I'm unsure of an easy way to recursively iterate through all the parents when a tile has been split more than once, like

```
get all locks for task 5 (with parent 3), 
since task 5 has a parent defined, get all locks for task 3 (with parent 1),
since task 3 has a parent defined, get all locks for task 1
```

without embedding a bunch of loops in Python. Some searches yielded things like CTE, joinedload, etc. in SQLAlchemy--none of which I'm sure are completely applicable.

pgiraud commented:
Hi @ethan-nelson 
I agree with you that it's not that easy to get the recursion working nicely. It's tricky though challenging. I wanted to have a try at it.
I didn't find a way to get this working with SQLAlchemy relationships. It looks like using CTE was the way to go though.

Here's the result of my work.
https://github.com/pgiraud/osm-tasking-manager2/commit/cdeedf9de23b447358ba54e5b59ee238b99efff7

Feel free to test it and go further.

ethan-nelson commented:
Still need to write tests, and for some reason I wasn't able to retain the ''committed with Pierre...''. Does this need an alembic upgrade?

ethan-nelson commented:
I wasn't able to retain the ''committed with Pierre'' so you can maybe do that locally.

Anyways, I went ahead and added an alembic upgrade code, but it could definitely use testing. Besides that, I added an additional constraint in the filter to not include `state_ready` as that unfortunately messes up the history when we include the initially created ready states after each split. Since that is added, we can get rid of the initial state pop in that list. 

I also removed the foreign key constraint on the parent_id. It unfortunately made things fairly difficult when using alembic and I don't think it's actually necessary (again my lack of RDBMS experience is showing).

Sampling:

![task_history](https://cloud.githubusercontent.com/assets/8998918/15169915/3210c070-1706-11e6-8a4b-756ea2b476b2.PNG)

pgiraud commented:
@ethan-nelson your explanations look good. I was about to propose the exit same adjustments.
The filtering on state_ready makes a lot of sense. And the foreign key is I agree not really necessary.

In the future, we could improve the look of the history by hiding the lock/unlock statements and simply keep "locked for xxx minutes" or "marked as done after xxxx minutes".

I'll give the code a look as soon as possible.

ethan-nelson commented:
Alright, great! And to your last point, I think there was an issue opened about that a year or two that everyone agreed basically to that effect (#503), but never got around to finishing.

pgiraud commented on alembic/versions/5002e75c0604_add_task_parent_to_retain_history_after_.py:
```diff
@@ -0,0 +1,21 @@
+def upgrade():
+    op.add_column('task', sa.Column('parent_id', sa.Integer(), nullable=True))
+
+def downgrade():
```
The downgrade script should remove the added column.

pgiraud replied:
In addition to this, I think it would be fairly easy to update the data for the already existing project. Guessing the parent task id by comparing the x, y for tiles with a different zoom from the one of the project is almost straight forward in my opinion. I can give a hand on this if you wish.

ethan-nelson replied:
I probably won't have time to tackle that for a few days, so if you want to go ahead and work on that you are welcome to. Though I do agree it shouldn't be too difficult to do.

pgiraud commented on osmtm/templates/task.history.mako:
```diff
@@ -52,6 +52,8 @@ from osmtm.mako_filters import (
         % elif step.state == step.state_validated:
           <span><i class="glyphicon glyphicon-thumbs-up text-success"></i> <b>${_('Validated')}</b> ${_('by')} ${user_link | n}</span>
+        % elif step.state == step.state_removed:
+          <span><i class="glyphicon icon-split"></i> <b>${_('Split')}</b> ${_('by')} ${user_link | n}</span>
```
By reading the translation here I just figured out that this may not work in some languages.
This should be addressed in a different issue but it would be better to have something looking like:
```
${_('Split by ${link}', mapping={'link': user_link})}
```

pgiraud commented:
Hey @ethan-nelson I finally managed to give the migration of data a try.
https://github.com/hotosm/osm-tasking-manager2/compare/master...pgiraud:retain-split-hist

It's not pretty but does the job. Takes several seconds to work on a database with many projects but works.

I also fixed an issue with markdown not correctly interpreted. (Added " | n ").
And filtered the locks to avoid the display of the first unlock entry when tasks are created.

Please give it a try, and cherry-pick the commits into your branch if you're happy.

ethan-nelson commented:
Alright I tested the script on a fresh db with some test task splits using the master branch, then switched to your remote branch and the db update worked great (not that I didn't expect it to)! There was a minor flake8 complaint I fixed and amended.

pgiraud commented:
Thanks for the test and the flake8 fix. Merging.
This was a nice addition. Hooray!

ethan-nelson commented on osmtm/views/task.py:
```diff
@@ -107,21 +134,21 @@ def task_xhr(request):
     locks = DBSession.query(TaskLock).filter(filter) \
         .order_by(TaskLock.date).all()
-    # remove the first lock (task creation)
-    locks.pop(0)
```
On the dev instance, I am seeing 'unlocked' for all tiles at the time of their creation. I think this might need to be retained.

pgiraud replied:
Indeed. "Task.lock is not None" isn't correct.
Pull request to come.

ethan-nelson replied:
Yeah, I guess I didn't check things after the flake8 fix...sorry for the error!
